src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_cipher.c

```diff
@@ -0,0 +1,328 @@
+DEFINE_CIPHER(Des3Ecb,      168, "DESede/ECB/NoPadding", false)
+DEFINE_CIPHER(Des3Cbc,      168, "DESede/CBC/NoPadding", false)
+DEFINE_CIPHER(Des3Cfb8,     168, "DESede/CFB/NoPadding", false)
+DEFINE_CIPHER(Des3Cfb64,    168, "DESede/CFB/NoPadding", false)
+DEFINE_UNSUPPORTED_CIPHER(RC2Ecb)
+DEFINE_UNSUPPORTED_CIPHER(RC2Cbc)
+
+
+static int32_t GetAlgorithmWidth(CipherInfo* type)
+{
+    if (!type->isSupported)
+    {
+        assert(false);
+        return FAIL;
+    }
+    return type->width;
+}
+
+static jobject GetAlgorithmName(JNIEnv* env, CipherInfo* type)
+{
+    if (!type->isSupported)
+    {
+        LOG_ERROR("This cipher is not supported");
+        assert(false);
+        return FAIL;
+    }
+    return JSTRING(type->name);
+}
+
+static bool HasTag(CipherInfo* type)
+{
+    return type->hasTag;
+}
+
+CipherCtx* AndroidCryptoNative_CipherCreatePartial(CipherInfo* type)
+{
+    JNIEnv* env = GetJNIEnv();
+    jobject algName = GetAlgorithmName(env, type);
+    if (!algName)
+        return FAIL;
+
+    jobject cipher = ToGRef(env, (*env)->CallStaticObjectMethod(env, g_cipherClass, g_cipherGetInstanceMethod, algName));
+    (*env)->DeleteLocalRef(env, algName);
+
+    if (CheckJNIExceptions(env))
+    {
+        return FAIL;
+    }
+
+    CipherCtx* ctx = malloc(sizeof(CipherCtx));
+    ctx->cipher = cipher;
+    ctx->type = type;
+    ctx->tagLength = TAG_MAX_LENGTH;
+    ctx->ivLength = 0;
+    ctx->encMode = 0;
+    ctx->key = NULL;
+    ctx->iv = NULL;
+    return ctx;
+}
+
+int32_t AndroidCryptoNative_CipherSetTagLength(CipherCtx* ctx, int32_t tagLength)
+{
+    if (!ctx)
+        return FAIL;
+    
+    if(tagLength > TAG_MAX_LENGTH)
+        return FAIL;
+
+    ctx->tagLength = tagLength;
+    return SUCCESS;
+}
+
+static int32_t ReinitializeCipher(CipherCtx* ctx)
+{
+    JNIEnv* env = GetJNIEnv();
+    
+    int32_t keyLength = GetAlgorithmWidth(ctx->type);
+
+    // int ivSize = cipher.getBlockSize();
+    // SecretKeySpec keySpec = new SecretKeySpec(key.getEncoded(), "AES");
+    // IvParameterSpec ivSpec = new IvParameterSpec(IV); or GCMParameterSpec for GCM/CCM
+    // cipher.init(encMode, keySpec, ivSpec);
+
+    jobject algName = GetAlgorithmName(env, ctx->type);
+    if (!algName)
+        return FAIL;
+
+    if (!ctx->ivLength)
+        ctx->ivLength = (*env)->CallIntMethod(env, ctx->cipher, g_getBlockSizeMethod);
+
+    jbyteArray keyBytes = (*env)->NewByteArray(env, keyLength / 8); // bits to bytes, e.g. 256 -> 32
+    (*env)->SetByteArrayRegion(env, keyBytes, 0, keyLength / 8, (jbyte*)ctx->key);
+    jbyteArray ivBytes = (*env)->NewByteArray(env, ctx->ivLength);
+    (*env)->SetByteArrayRegion(env, ivBytes, 0, ctx->ivLength, (jbyte*)ctx->iv);
+
+    jobject sksObj = (*env)->NewObject(env, g_sksClass, g_sksCtor, keyBytes, algName);
+    jobject ivPsObj;
+
+    if (HasTag(ctx->type))
+        ivPsObj = (*env)->NewObject(env, g_GCMParameterSpecClass, g_GCMParameterSpecCtor, ctx->tagLength * 8, ivBytes);
```

Comment:
Do both GCM and CCM use the GCMParameterSpec class?

Reply:
Yes they both use the same class.